network/tools/network-policy-enforcement-latency/pkg/client.go

```diff
@@ -0,0 +1,552 @@
+	// The number of workers to process the pod watch events in the workerQueue.
+	workerCount int
+}
+
+type boolMapWithLock struct {
+	mp   map[string]bool
+	lock sync.RWMutex
+}
+
+type timeMapWithLock struct {
+	mp   map[string]time.Time
+	lock sync.RWMutex
+}
+
+type timeWithLock struct {
+	time *time.Time
+	lock sync.RWMutex
+}
+
+type targetSpec struct {
+	ip, name, namespace string
+	startTime           time.Time
+}
+
+type podEvent struct {
+	podName    string
+	isAddEvent bool
+}
+
+// CreateConfig creates a configuration object for test client based on the
+// command flags.
+func CreateConfig() *config {
+	config := config{}
+
+	flag.StringVar(&config.testClientNamespace, "testClientNamespace", "", "The namespace of test client pods")
+	flag.StringVar(&config.targetLabelSelector, "targetLabelSelector", "", "The label selector of target pods to send requests to")
+	flag.StringVar(&config.targetNamespace, "targetNamespace", "", "The namespace of target pods to send requests to")
+	flag.IntVar(&config.targetPort, "targetPort", 0, "The port number of target pods to send requests to")
+	flag.IntVar(&config.maxTargets, "maxTargets", 100, "The maximum number of target pods to send requests to")
+	flag.StringVar(&config.allowPolicyName, "allowPolicyName", "", "The name of the egress policy that allows traffic from test client pods to target pods")
+	flag.IntVar(&config.metricsPort, "metricsPort", 9154, "The port number where Prometheus metrics are exposed")
+	flag.BoolVar(&config.podCreationMode, "podCreationMode", false, "Specifies whether to test latency for pod creation mode")
+	flag.IntVar(&config.workerCount, "workerCount", utils.DefaultWorkerCount, "The number of workers to process the pod watch events in the workerQueue")
+	flag.Parse()
+
+	return &config
+}
+
+// NewTestClient creates a new test client for the provided config.
+func NewTestClient(config *config, mainStopChan chan os.Signal) *TestClient {
+	return &TestClient{
+		config:       config,
+		MainStopChan: mainStopChan,
+	}
+}
+
+// Run verifies the test client config, starts the metrics server, and runs the
+// network policy enforcement latency test based on the config.
+func (c *TestClient) Run() {
+	if err := c.initialize(); err != nil {
+		klog.Errorf("Failed to initialize: %v", err)
+		c.enterIdleState()
+		return
+	}
+
+	c.metricsServer = metrics.StartMetricsServer(fmt.Sprintf(":%d", c.config.metricsPort))
+	metrics.RegisterMetrics(c.config.podCreationMode)
+	defer func(metricsServer *http.Server, ctx context.Context) {
+		if err := metricsServer.Shutdown(ctx); err != nil {
+			klog.Errorf("Metrics server shutdown failed, error: %v", err)
+		}
+	}(c.metricsServer, context.TODO())
+
+	if c.config.podCreationMode {
+		if err := c.measurePodCreation(); err != nil {
+			klog.Errorf("Pod creation test failed, error: %v", err)
+		}
+	} else {
+		if err := c.measureNetPolicyCreation(); err != nil {
+			klog.Errorf("Pod creation test failed, error: %v", err)
+		}
+	}
```

Comment:
I don't like it. If I'm not mistake majority of the code is for either pod or policy.
Some common code could be extracted to some utils or library.

Reply:
You are right. Most of the functionality besides initialization and sending a request is not common to the modes.
I will extract the common code to utils package and separate it into 2 applications. We can run both in the scale tests.
Sounds good?